### PR TITLE
GPU jpeg decoder: add batch support and hardware decoding

### PR DESCRIPTION
**Pull request by deekay42**
Over 8000 imgs/s on 1 A100 GPU
```

Platform: Linux-5.12.0-0_fbk7_zion_6511_gd766966f605a-x86_64-with-glibc2.34
Logical CPUs: 192

CUDA device: NVIDIA PG509-210
Total Memory: 84.99 GB

Mean image size: 551x676
[---------------------------------------------------------------- Image Decoding ----------------------------------------------------------------]
                                                                                                        |  1 images  |  100 images  |  1000 images
1 threads: ---------------------------------------------------------------------------------------------------------------------------------------
      CPU (unfused): [torchvision.io.decode_jpeg(img, device='cpu') for img in encoded_images_trunc]    |   3301.9   |   271141.6   |   2541465.3 
      CPU (fused): torchvision.io.decode_jpeg(encoded_images_trunc, device='cpu')                       |   3239.7   |   288522.8   |   2596394.3 
      CUDA (unfused): [torchvision.io.decode_jpeg(img, device='cuda') for img in encoded_images_trunc]  |    603.7   |    60097.8   |    573783.4 
      CUDA (fused): torchvision.io.decode_jpeg(encoded_images_trunc, device='cuda')                     |    600.6   |    12972.6   |    127654.8 
12 threads: --------------------------------------------------------------------------------------------------------------------------------------
      CPU (unfused): [torchvision.io.decode_jpeg(img, device='cpu') for img in encoded_images_trunc]    |   3330.5   |   272498.9   |   2552944.3 
      CPU (fused): torchvision.io.decode_jpeg(encoded_images_trunc, device='cpu')                       |   3339.7   |   257796.7   |   2511005.4 
      CUDA (unfused): [torchvision.io.decode_jpeg(img, device='cuda') for img in encoded_images_trunc]  |    603.8   |    59138.0   |    588341.4 
      CUDA (fused): torchvision.io.decode_jpeg(encoded_images_trunc, device='cuda')                     |    605.0   |    13163.7   |    127891.4 
24 threads: --------------------------------------------------------------------------------------------------------------------------------------
      CPU (unfused): [torchvision.io.decode_jpeg(img, device='cpu') for img in encoded_images_trunc]    |   3227.5   |   276357.8   |   2518914.3 
      CPU (fused): torchvision.io.decode_jpeg(encoded_images_trunc, device='cpu')                       |   3277.7   |   257554.9   |   2497894.3 
      CUDA (unfused): [torchvision.io.decode_jpeg(img, device='cuda') for img in encoded_images_trunc]  |    607.9   |    58306.1   |    583932.6 
      CUDA (fused): torchvision.io.decode_jpeg(encoded_images_trunc, device='cuda')                     |    653.2   |    12604.1   |    124130.5 

Times are in microseconds (us).```
```

### CODE REVIEW

File: torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h [hunk 1]
```diff
@@ -2,14 +2,15 @@
...
 #include "encode_jpegs_cuda.h"
 
 namespace vision {
 namespace image {
 
-C10_EXPORT torch::Tensor decode_jpeg_cuda(
-    const torch::Tensor& data,
-    ImageReadMode mode,
+C10_EXPORT std::vector<torch::Tensor> decode_jpegs_cuda(
+    const std::vector<torch::Tensor>& encoded_images,
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h:12]: Why is this not a single torch::Tensor that's stacked as opposed to a std::vector of Tensors?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+    const std::vector<torch::Tensor>& encoded_images,
+    vision::image::ImageReadMode mode,
+    torch::Device device) {
+  C10_LOG_API_USAGE_ONCE(
+      "torchvision.csrc.io.image.cuda.decode_jpegs_cuda.decode_jpegs_cuda");
+
+  std::lock_guard<std::mutex> lock(decoderMutex);
+
+  for (auto& encoded_image : encoded_images) {
+    TORCH_CHECK(
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:45]: Should you check for is_contiguous here too?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+        !encoded_image.is_cuda(),
+        "The input tensor must be on CPU when decoding with nvjpeg")
+
+    TORCH_CHECK(
+        encoded_image.dim() == 1 && encoded_image.numel() > 0,
+        "Expected a non empty 1-dimensional tensor");
+  }
+
+  TORCH_CHECK(device.is_cuda(), "Expected a cuda device");
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:57]: Move this above the for loop? Typically cheaper checks are done before more expensive ones

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+    TORCH_WARN_ONCE(
+        "There is a memory leak issue in the nvjpeg library for CUDA versions < 11.6. "
+        "Make sure to rely on CUDA 11.6 or above before using decode_jpeg(..., device='cuda').");
+  }
+
+  at::cuda::CUDAGuard device_guard(device);
+
+  if (cudaJpegDecoder == nullptr || device != cudaJpegDecoder->target_device) {
+    if (cudaJpegDecoder != nullptr)
+      delete cudaJpegDecoder.release();
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:84]: You can do cudaJpegDecoder.reset() instead of manually deleting

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+    cudaJpegDecoder = std::make_unique<CUDAJpegDecoder>(device);
+    std::atexit([]() { delete cudaJpegDecoder.release(); });
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:86]: Use .reset() here as well instead of manually deleting?

File: torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h [hunk 6]
```diff
@@ -2,14 +2,15 @@
...
+#include "decode_jpegs_cuda.h"
 #include "encode_jpegs_cuda.h"
 
 namespace vision {
 namespace image {
 
-C10_EXPORT torch::Tensor decode_jpeg_cuda(
-    const torch::Tensor& data,
-    ImageReadMode mode,
+C10_EXPORT std::vector<torch::Tensor> decode_jpegs_cuda(
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h:11]: Add a comment with example usage and comment about thread safety. Can multiple threads call this function concurrently?

Comment (ahmadsharif1) [reply torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h:11]: Also, what happens on errors? If a single image fails, will it fail the whole batch?

Does it decode everything in parallel?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+    default:
+      TORCH_CHECK(
+          false, "The provided mode is not supported for JPEG decoding on GPU");
+  }
+
+  try {
+    at::cuda::CUDAEvent event;
+    event.record(cudaJpegDecoder->stream);
+    auto result = cudaJpegDecoder->decode_images(encoded_images, output_format);
+    if (device.has_index())
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:114]: Style nit: add curly braces to the if cause it's multiline

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+      event.block(at::cuda::getCurrentCUDAStream(
+          cudaJpegDecoder->original_device.index()));
+    else
+      event.block(at::cuda::getCurrentCUDAStream());
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:118]: Nit: it may look more readable to have a variable for the stream (you can have an if condition for that). And then do a single event.block on that stream

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+      output_format = NVJPEG_OUTPUT_RGB;
+      break;
+    default:
+      TORCH_CHECK(
+          false, "The provided mode is not supported for JPEG decoding on GPU");
+  }
+
+  try {
+    at::cuda::CUDAEvent event;
+    event.record(cudaJpegDecoder->stream);
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:112]: I may be missing something, but shouldn't the record be after the work is done if you are going to wait for it to finish on L118?

I mean shouldn't it be after decode_images()?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 10]
```diff
@@ -0,0 +1,582 @@
...
+  status = nvjpegDecodeParamsCreate(nvjpeg_handle, &nvjpeg_decode_params);
+  TORCH_CHECK(
+      status == NVJPEG_STATUS_SUCCESS,
+      "Failed to create decode params: ",
+      status);
+}
+
+CUDAJpegDecoder::~CUDAJpegDecoder() {
+  /*
+  The below code works on Mac and Linux, but fails on Windows.
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:222]: Do you want to use a #ifdef _WIN32 here?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.h [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+  std::vector<torch::Tensor> decode_images(
+      const std::vector<torch::Tensor>& encoded_images,
+      const nvjpegOutputFormat_t& output_format);
+
+  const torch::Device original_device;
+  const torch::Device target_device;
+  const c10::cuda::CUDAStream stream;
+
+ protected:
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.h:25]: Out of curiosity, why protected and not private?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 12]
```diff
@@ -0,0 +1,582 @@
...
+    containing the decoded images
+  */
+
+  auto [decoded_imgs_buf, output_tensors, channels] =
+      prepare_buffers(encoded_images, output_format);
+
+  nvjpegStatus_t status;
+  cudaError_t cudaStatus;
+
+  cudaStatus = cudaStreamSynchronize(stream);
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:405]: Why is this needed?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 13]
```diff
@@ -0,0 +1,582 @@
...
+          &sw_output_buffer[i],
+          stream);
+      TORCH_CHECK(
+          status == NVJPEG_STATUS_SUCCESS,
+          "Failed to decode jpeg stream: ",
+          status);
+    }
+  }
+
+  cudaStatus = cudaStreamSynchronize(stream);
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:563]: Why do you have the outer CUDAEvent in the caller of this function if you already wait for all ops to finish here?

**Review by ahmadsharif1 — COMMENTED** [review]
Flushing out some comments on the C++ code.

**Review by ahmadsharif1 — COMMENTED** [review]
Flushing out some comments on the C++ code.

File: torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h (hunk 1 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h:12]: Because the encoded jpegs are represented as variable length byte streams of type tensor(1). They cannot be stacked into a batch size they don't have the same dimensions.

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp (hunk 2 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:45]: Seems unlikely but I guess technically even ]tensors of shape (1,) could be non-contiguous. I'll add some conversion code.

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp (hunk 3 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:57]: This needs to happen for every tensor in the vector though. Users could pass a list with tensors like [cuda, cpu, cuda]

File: torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h (hunk 6 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h:11]: The C++ functions are primarily meant to be used by the PyBind code only. I didn't really intend for them to be used by libtorch users, but I suppose there isn't really any harm in it either. Added docstrings.

Comment (deekay42) [reply torchvision/csrc/io/image/cuda/encode_decode_jpegs_cuda.h:11]: To summarize: 
- Yes, both functions are thread-safe.
- Errors in one image fails the whole function. No partial results are returned.
- The decoding happens in parallel

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp (hunk 9 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:112]: Good catch! Of course event.record should come after the work is done.

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp (hunk 10 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:222]: I've thought about it but I understand that C++ order of destruction is generally undefined, so even if it passes on the specific Mac and Linux versions I've tested it on, it still is undefined behavior and may fail at any time.

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.h (hunk 11 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.h:25]: No real reason. Happy to change to private

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp (hunk 12 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:405]: That's how they do it here: https://github.com/NVIDIA/CUDALibrarySamples/blob/f17940ac4e705bf47a8c39f5365925c1665f6c98/nvJPEG/nvJPEG-Decoder/nvjpegDecoder.cpp#L36
After buffers are allocated they synchronize before starting the decoding

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp (hunk 13 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:563]: Line 563 is needed because I need to do some pruning right after. The outer CUDAEvent is needed to synchronize the internal CUDAJpegDecoder::stream with the calling code's current stream before returning the results

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 14]
```diff
@@ -0,0 +1,592 @@
...
+      cudaStatus == cudaSuccess,
+      "Failed to synchronize CUDA stream: ",
+      cudaStatus);
+
+  // prune extraneous channels from single channel images
+  if (output_format == NVJPEG_OUTPUT_UNCHANGED)
+    for (std::vector<at::Tensor>::size_type i = 0; i < output_tensors.size();
+         ++i)
+      if (channels[i] == 1)
+        output_tensors[i] = output_tensors[i][0].unsqueeze(0).clone();
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:584]: Out of curiosity -- why is is clone needed here? clone could be expensive hence my question

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 15]
```diff
@@ -0,0 +1,592 @@
...
+
+    Returns:
+    - decoded_images (std::vector<nvjpegImage_t>): a vector of nvjpegImages
+    containing pointers to the memory of the decoded images
+    - output_tensors (std::vector<torch::Tensor>): a vector of Tensors
+    containing the decoded images. `decoded_images` points to the memory of
+    output_tensors
+    - channels (std::vector<int>): a vector of ints containing the number of
+    output image channels for every image
+  */
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:329]: Move this to the header file?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.h [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+      const nvjpegOutputFormat_t& output_format);
+  nvjpegJpegState_t nvjpeg_state;
+  nvjpegJpegState_t nvjpeg_decoupled_state;
+  nvjpegBufferPinned_t pinned_buffers[2];
+  nvjpegBufferDevice_t device_buffer;
+  nvjpegJpegStream_t jpeg_streams[2];
+  nvjpegDecodeParams_t nvjpeg_decode_params;
+  nvjpegJpegDecoder_t nvjpeg_decoder;
+  bool hw_decode_available{true};
+  nvjpegHandle_t nvjpeg_handle;
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.h:41]: Style nit: Class member usually have suffixes of "_" to separate them from local variables.

https://google.github.io/styleguide/cppguide.html#Variable_Names

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 17]
```diff
@@ -0,0 +1,592 @@
...
+
+  try {
+    at::cuda::CUDAEvent event;
+    auto result = cudaJpegDecoder->decode_images(contig_images, output_format);
+    auto current_stream{
+        device.has_index() ? at::cuda::getCurrentCUDAStream(
+                                 cudaJpegDecoder->original_device.index())
+                           : at::cuda::getCurrentCUDAStream()};
+    event.record(cudaJpegDecoder->stream);
+    event.block(current_stream) return result;
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:129]: Why is there no semi-colon after ")" here? Does this code compile properly?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 18]
```diff
@@ -0,0 +1,592 @@
...
+    This function scans the encoded images' jpeg headers and
+    allocates decoding buffers based on the metadata found
+
+    Args:
+    - encoded_images (std::vector<torch::Tensor>): a vector of tensors
+    containing the jpeg bitstreams to be decoded. Each tensor must have dtype
+    torch.uint8 and device cpu
+    - output_format (nvjpegOutputFormat_t): NVJPEG_OUTPUT_RGB, NVJPEG_OUTPUT_Y
+    or NVJPEG_OUTPUT_UNCHANGED
+    - device (torch::Device): The desired CUDA device for the returned Tensors
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:319]: This is a stale comment since there is no device arg

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 19]
```diff
@@ -0,0 +1,592 @@
...
+
+  nvjpegOutputFormat_t output_format;
+
+  switch (mode) {
+    case vision::image::IMAGE_READ_MODE_UNCHANGED:
+      // Using NVJPEG_OUTPUT_UNCHANGED causes differently sized output channels
+      // which is related to the subsampling used I'm not sure why this is the
+      // case, but for now we're just using RGB and later removing channels from
+      // grayscale images.
+      output_format = NVJPEG_OUTPUT_UNCHANGED;
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:108]: Nit: would it be simpler to just use NVJPEG_OUTPUT_RGB here since you are assuming this expands the channels anyway?

Also add a TODO to investigate and fix this behavior of pruning

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 20]
```diff
@@ -0,0 +1,592 @@
...
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <typeinfo>
+namespace vision {
+namespace image {
+
+std::mutex decoderMutex;
+std::unique_ptr<CUDAJpegDecoder> cudaJpegDecoder;
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:33]: Nit: maybe use gCudaJpegDecoder to indicate it's a global variable?

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 21]
```diff
@@ -0,0 +1,592 @@
...
+
+CUDAJpegDecoder::~CUDAJpegDecoder() {
+  /*
+  The below code works on Mac and Linux, but fails on Windows.
+  This is because on Windows, the atexit hook which calls this
+  destructor executes after cuda is already shut down causing SIGSEGV.
+  We do not have a solution to this problem at the moment, so we'll
+  just leak the libnvjpeg & cuda variables for the time being and hope
+  that the CUDA runtime handles cleanup for us.
+  Please send a PR if you have a solution for this problem.
```
Comment (ahmadsharif1) [inline torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:238]: One request: maybe try the driver API to see if cuda is available?

https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__DEVICE.html#group__CUDA__DEVICE_1g52b5ce05cb8c5fb6831b2c0ff2887c74

int dummy;
if (cuDeviceGetCount (&dummy) == CUDA_SUCESS) {
   ...
 }

If that doesn't work don't bother with a unique_ptr for the global variable and just use a regular pointer.

Add a comment above the global variable saying this is not a unique_ptr because our destructor could race with cuda's own destructors

File: benchmarks/encoding_decoding.py [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+    for device in ["cpu", "cuda"]:
+        decoded_images_device = [t.to(device=device) for t in decoded_images]
+        for size in [1, 100, 1000]:
+            for num_threads in [1, 12, 24]:
+                for stmt, strat in zip(
+                    [
+                        "[torchvision.io.encode_jpeg(img) for img in decoded_images_device_trunc]",
+                        "torchvision.io.encode_jpeg(decoded_images_device_trunc)",
+                    ],
+                    ["unfused", "fused"],
```
Comment (ahmadsharif1) [inline benchmarks/encoding_decoding.py:45]: I could be wrong, but batched seems like a better term than fused since it appears to be batching images, not fusing kernels necessarily.

**Review by NicolasHug — APPROVED** [review]
Thanks a ton @deekay42

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp [hunk 23]
```diff
@@ -0,0 +1,592 @@
...
+          nvjpeg_handle,
+          nvjpeg_decoder,
+          nvjpeg_decoupled_state,
+          &sw_output_buffer[i],
+          stream);
+      TORCH_CHECK(
+          status == NVJPEG_STATUS_SUCCESS,
+          "Failed to decode jpeg stream: ",
+          status);
+    }
```
[the remark this answers is not in the document]
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:575]: There are many different types of jpegs, for out purposes most notably baseline and progressive. The main difference between the two is that progressive jpegs encapsulate multiple renderings of the same image at different resolutions. Baseline jpegs can be decoded on the GPU right away, but progressive jpegs need some preprocessing on the host before the GPU can process them. Added a comment

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp (hunk 14 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:589]: To be quite honest, I don't remember and at this point I'll probably just go ahead and land this and we can send a follow up PR if needed. In any event, this would only apply to grayscale images

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.h (hunk 16 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.h:41]: Will keep that in mind for next time!

File: torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp (hunk 21 above)
Comment (deekay42) [reply torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp:240]: Yea I tried that and it didn't work :(

File: benchmarks/encoding_decoding.py (hunk 22 above)
Comment (deekay42) [reply benchmarks/encoding_decoding.py:45]: If the images are batched it uses a fused kernel
